File: android/sdk/src/main/java/com/viant/forgeandroid/runtime/WorkflowPrimitiveRuntime.kt
package com.viant.forgeandroid.runtime

enum class DataStateBoundaryKind {
    Loading,
    Empty,
    Partial,
    Error,
    Ready
}

data class HistoryDiffEntry(val path: String, val label: String, val before: String, val after: String, val redacted: Boolean)
data class ScheduleValidationError(val index: Int, val code: String)
data class ScheduleValidationResult(val valid: Boolean, val errors: List<ScheduleValidationError>)
data class UploadFileValue(val name: String, val mimeType: String = "application/octet-stream", val data: ByteArray)
data class UploadValidationResult(val valid: Boolean, val errors: List<String>)

object WorkflowPrimitiveRuntime {
    fun dataStateBoundaryKind(
        controls: List<ControlState>,
        collections: List<List<Map<String, Any?>>>,
        allowPartial: Boolean = false
    ): DataStateBoundaryKind {
        val hasError = controls.any { !it.error.isNullOrBlank() }
        val loading = controls.any { it.loading }
        val populated = collections.count { it.isNotEmpty() }
        val sourceCount = maxOf(controls.size, collections.size)

        if (loading && populated == 0) return DataStateBoundaryKind.Loading
        if (hasError) {
            return if (allowPartial && populated > 0) DataStateBoundaryKind.Partial else DataStateBoundaryKind.Error
        }
        if (sourceCount > 1 && populated in 1 until sourceCount) {
            return if (allowPartial) DataStateBoundaryKind.Partial else DataStateBoundaryKind.Empty
        }
        return if (populated == 0) DataStateBoundaryKind.Empty else DataStateBoundaryKind.Ready
    }

    fun presentationRecord(
        form: Map<String, Any?>,
        collection: List<Map<String, Any?>>,
        metrics: Map<String, Any?>
    ): Map<String, Any?> = when {
        form.isNotEmpty() -> form
        collection.isNotEmpty() -> collection.first()
        else -> metrics
    }

    fun relationLabel(count: Int, spec: RelationDrillSpec): String {
        if (count == 0 && !spec.emptyText.isNullOrBlank()) return spec.emptyText
        val noun = if (count == 1) spec.singularLabel ?: "item" else spec.pluralLabel ?: "items"
        return "$count $noun"
    }

    fun permissionAllows(
        spec: PermissionBoundarySpec,
        authorization: Map<String, Any?>,
        rows: List<Map<String, Any?>> = emptyList(),
        grants: List<Map<String, Any?>> = emptyList()
    ): Boolean {
        val capability = spec.capability?.trim().orEmpty()
        if (capability.isEmpty()) return true
        val mode = spec.mode?.trim()?.lowercase() ?: "resource"
        if (mode == "resource") {
            val resource = JsonUtil.asStringMap(authorization["resource"])
            return JsonUtil.asStringMap(resource["capabilities"])[capability] == true
        }
        if (rows.isEmpty()) return true
        val identity = spec.identityField?.trim().orEmpty().ifBlank { "id" }
        val grantsById = grants.mapNotNull { grant ->
            (grant["resourceId"] ?: grant[identity])?.toString()?.takeIf(String::isNotBlank)?.let { it to grant }
        }.toMap()
        return rows.all { row ->
            val key = row[identity]?.toString()?.takeIf(String::isNotBlank) ?: return@all false
            val grant = grantsById[key] ?: return@all false
            JsonUtil.asStringMap(grant["capabilities"])[capability] == true
        }
    }

    fun responsiveDataGridState(spec: ResponsiveDataGridSpec, target: String): ResponsiveDataGridStateSpec? =
        spec.breakpoints[target]
            ?: spec.breakpoints[if (target == "phone") "narrow" else "wide"]
            ?: spec.breakpoints["desktop"]
            ?: spec.breakpoints.values.firstOrNull()

    fun validateSchedule(rows: List<Map<String, Any?>>, spec: ScheduleEditorSpec): ScheduleValidationResult {
        val startField = spec.startField ?: "start"
        val endField = spec.endField ?: "end"
        val minimum = parseDurationMillis(spec.minDuration)
        val normalized = mutableListOf<Triple<Int, java.time.Instant?, java.time.Instant?>>()
        val errors = mutableListOf<ScheduleValidationError>()
        rows.forEachIndexed { index, row ->
            if (JsonUtil.asStringMap(row["_scheduleErrors"]).values.any { it == true || it?.toString()?.isNotBlank() == true }) errors += ScheduleValidationError(index, "timezone")
            val start = scheduleInstant(SelectorUtil.resolve(row, startField))
            val end = scheduleInstant(SelectorUtil.resolve(row, endField))
            normalized += Triple(index, start, end)
            if (start == null || end == null) errors += ScheduleValidationError(index, "invalid_date")
            else {
                if (!end.isAfter(start)) errors += ScheduleValidationError(index, "invalid_range")
                if (minimum != null && java.time.Duration.between(start, end).toMillis() < minimum) errors += ScheduleValidationError(index, "min_duration")
            }
        }
        if (spec.allowOverlap == false) {
            val sorted = normalized.filter { it.second != null && it.third != null }.sortedBy { it.second }
            for (index in 1 until sorted.size) if (sorted[index].second!!.isBefore(sorted[index - 1].third!!)) errors += ScheduleValidationError(sorted[index].first, "overlap")
        }
        return ScheduleValidationResult(errors.isEmpty(), errors)
    }

    fun toggleTreeSelection(nodes: List<Map<String, Any?>>, selected: Set<String>, key: String, checked: Boolean, spec: TreeEditorSpec): Set<String> {
        val keys = mutableSetOf(key)
        if (spec.cascade.equals("descendants", true)) nodes.forEach { collectTreeKeys(it, key, false, spec, keys) }
        return selected.toMutableSet().apply { keys.forEach { if (checked) add(it) else remove(it) } }
    }

    fun validateUpload(files: List<UploadFileValue>, spec: UploadCollectionSpec): UploadValidationResult {
        val errors = mutableListOf<String>()
        if (spec.multiple == false && files.size > 1) errors += "Only one file is allowed."
        if ((spec.maxFiles ?: 0) > 0 && files.size > spec.maxFiles!!) errors += "No more than ${spec.maxFiles} files are allowed."
        files.forEach { file ->
            if (!uploadTypeAccepted(file, spec.accept)) errors += "${file.name} has an unsupported type."
            if ((spec.maxBytes ?: 0) > 0 && file.data.size > spec.maxBytes!!) errors += "${file.name} exceeds the size limit."
        }
        return UploadValidationResult(errors.isEmpty(), errors)
    }

    fun encodeUpload(files: List<UploadFileValue>, spec: UploadCollectionSpec): Map<String, Any?> {
        val transport = spec.transport ?: "mcpBlob"
        require(transport == "mcpBlob") { "Unsupported upload transport: $transport" }
        val blobs = files.map { file -> mapOf(
            "name" to file.name, "size" to file.data.size, "type" to file.mimeType,
            "data" to java.util.Base64.getEncoder().encodeToString(file.data), "filename" to file.name, "mimeType" to file.mimeType
        ) }
        return mapOf("transport" to "mcpBlob", (spec.blobField ?: "files") to if (spec.multiple == false) blobs.firstOrNull() else blobs)
    }

    fun masterDetailIdentity(row: Map<String, Any?>?, fields: List<String>): String? {
        if (row == null) return null
        val keys = fields.ifEmpty { listOf("id") }
        val values = keys.map { row[it]?.toString()?.takeIf(String::isNotBlank) ?: return null }
        return values.joinToString("\u001f")
    }

    fun resolveMasterDetailSelection(rows: List<Map<String, Any?>>, selected: Map<String, Any?>?, persisted: Map<String, Any?>?, identityFields: List<String>): Map<String, Any?>? {
        masterDetailIdentity(selected, identityFields)?.let { id -> rows.firstOrNull { masterDetailIdentity(it, identityFields) == id }?.let { return it } }
        masterDetailIdentity(persisted, identityFields)?.let { id -> return rows.firstOrNull { masterDetailIdentity(it, identityFields) == id } }
        return null
    }

    fun masterDetailParameters(spec: MasterDetailRegionSpec, row: Map<String, Any?>): Map<String, Any?> = spec.parameters.mapValues { (_, parameter) ->
        val objectValue = parameter as? kotlinx.serialization.json.JsonObject ?: return@mapValues JsonUtil.elementToAny(parameter)
        val source = objectValue["source"]?.let(JsonUtil::elementToAny)?.toString()?.lowercase() ?: "row"
        if (source != "row") return@mapValues objectValue["value"]?.let(JsonUtil::elementToAny)
        val selector = (objectValue["selector"] ?: objectValue["field"])?.let(JsonUtil::elementToAny)?.toString().orEmpty()
        val value = SelectorUtil.resolve(row, selector)
        if (objectValue["wrap"]?.let(JsonUtil::elementToAny)?.toString()?.lowercase() == "array") listOf(value) else value
    }

    private fun collectTreeKeys(node: Map<String, Any?>, target: String, collecting: Boolean, spec: TreeEditorSpec, result: MutableSet<String>) {
        val nodeKey = SelectorUtil.resolve(node, spec.identityField ?: "id")?.toString().orEmpty()
        val shouldCollect = collecting || nodeKey == target
        if (shouldCollect && nodeKey.isNotEmpty()) result += nodeKey
        @Suppress("UNCHECKED_CAST")
        val children = SelectorUtil.resolve(node, spec.childrenField ?: "children") as? List<Map<String, Any?>> ?: emptyList()
        children.forEach { collectTreeKeys(it, target, shouldCollect, spec, result) }
    }

    fun historyDiffEntries(before: Any?, after: Any?, spec: HistoryDiffSpec): List<HistoryDiffEntry> {
        val result = mutableListOf<HistoryDiffEntry>()
        collectHistoryDiff(before, after, "", spec, result)
        return result
    }

    private fun collectHistoryDiff(before: Any?, after: Any?, path: String, spec: HistoryDiffSpec, result: MutableList<HistoryDiffEntry>) {
        val leaf = path.substringAfterLast('.', path)
        if (path in spec.ignoreFields || leaf in spec.ignoreFields) return
        val left = before as? Map<*, *>
        val right = after as? Map<*, *>
        if (left != null && right != null) {
            (left.keys + right.keys).filterIsInstance<String>().toSortedSet().forEach { key ->
                collectHistoryDiff(left[key], right[key], if (path.isEmpty()) key else "$path.$key", spec, result)
            }
            return
        }
        val normalizedBefore = normalizeHistoryValue(before, spec.arrayStrategy)
        val normalizedAfter = normalizeHistoryValue(after, spec.arrayStrategy)
        if (normalizedBefore == normalizedAfter) return
        val redacted = path in spec.redactFields || leaf in spec.redactFields
        result += HistoryDiffEntry(
            path.ifEmpty { "value" },
            spec.fieldLabels[path] ?: spec.fieldLabels[leaf] ?: historyLabel(leaf),
            if (redacted) "••••" else historyDisplay(before),
            if (redacted) "••••" else historyDisplay(after),
            redacted
        )
    }

    private fun normalizeHistoryValue(value: Any?, strategy: String?): Any? =
        if (strategy.equals("set", true) && value is List<*>) value.map(::historyDisplay).sorted() else value

    private fun historyDisplay(value: Any?): String = when (value) {
        null -> "—"
        is Map<*, *> -> value.entries.sortedBy { it.key.toString() }.joinToString(", ") { "${it.key}: ${historyDisplay(it.value)}" }
        is Iterable<*> -> value.joinToString(", ") { historyDisplay(it) }
        is Number -> if (value.toDouble() % 1.0 == 0.0) value.toLong().toString() else value.toString()
        else -> value.toString()
    }

    private fun historyLabel(value: String): String = value.replace('_', ' ').split(' ').joinToString(" ") { word -> word.replaceFirstChar { it.uppercase() } }
}

private fun uploadTypeAccepted(file: UploadFileValue, rules: List<String>): Boolean {
    if (rules.isEmpty()) return true
    val name = file.name.lowercase(); val type = file.mimeType.lowercase()
    return rules.any { raw ->
        val rule = raw.lowercase()
        when { rule.startsWith(".") -> name.endsWith(rule); rule.endsWith("/*") -> type.startsWith(rule.dropLast(1)); else -> type == rule }
    }
}

private fun scheduleInstant(value: Any?): java.time.Instant? = runCatching {
    when (value) {
        is java.time.Instant -> value
        is String -> java.time.Instant.parse(value)
        else -> null
    }
}.getOrNull()

private fun parseDurationMillis(raw: String?): Long? {
    val match = Regex("^(\\d+(?:\\.\\d+)?)\\s*(ms|s|m|h|d)$", RegexOption.IGNORE_CASE).matchEntire(raw?.trim().orEmpty()) ?: return null
    val value = match.groupValues[1].toDoubleOrNull() ?: return null
    val multiplier = when (match.groupValues[2].lowercase()) { "ms" -> 1.0; "s" -> 1_000.0; "m" -> 60_000.0; "h" -> 3_600_000.0; "d" -> 86_400_000.0; else -> 1.0 }
    return (value * multiplier).toLong()
}

File: ios/Sources/ForgeIOSRuntime/WorkflowPrimitiveRuntime.swift
import Foundation

public enum DataStateBoundaryKind: String, Sendable, Equatable {
    case loading
    case empty
    case partial
    case error
    case ready
}

public struct HistoryDiffEntry: Sendable, Equatable, Identifiable {
    public let path: String
    public let label: String
    public let before: String
    public let after: String
    public let redacted: Bool
    public var id: String { path }
}

public struct ScheduleValidationError: Sendable, Equatable {
    public let index: Int
    public let code: String
}

public struct ScheduleValidationResult: Sendable, Equatable {
    public let valid: Bool
    public let errors: [ScheduleValidationError]
}

public struct UploadFileValue: Sendable, Equatable, Identifiable {
    public let name: String
    public let mimeType: String
    public let data: Data
    public var id: String { "\(name):\(data.count)" }

    public init(name: String, mimeType: String = "application/octet-stream", data: Data) {
        self.name = name
        self.mimeType = mimeType
        self.data = data
    }
}

public struct UploadValidationResult: Sendable, Equatable {
    public let valid: Bool
    public let errors: [String]
}

public enum WorkflowPrimitiveRuntime {
    public static func dataStateBoundaryKind(
        controls: [ControlState],
        collections: [[[String: JSONValue]]],
        allowPartial: Bool = false
    ) -> DataStateBoundaryKind {
        let errors = controls.filter { ($0.error ?? "").isEmpty == false }
        let loading = controls.contains { $0.loading }
        let populated = collections.filter { !$0.isEmpty }.count
        let sourceCount = max(controls.count, collections.count)

        if loading && populated == 0 { return .loading }
        if !errors.isEmpty {
            if allowPartial && populated > 0 { return .partial }
            return .error
        }
        if sourceCount > 1 && populated > 0 && populated < sourceCount {
            return allowPartial ? .partial : .empty
        }
        return populated == 0 ? .empty : .ready
    }

    public static func presentationRecord(
        form: [String: JSONValue],
        collection: [[String: JSONValue]],
        metrics: [String: JSONValue]
    ) -> [String: JSONValue] {
        if !form.isEmpty { return form }
        if let first = collection.first { return first }
        return metrics
    }

    public static func relationLabel(count: Int, spec: RelationDrillSpec) -> String {
        if count == 0, let empty = spec.emptyText, !empty.isEmpty { return empty }
        let noun = count == 1
            ? (spec.singularLabel ?? "item")
            : (spec.pluralLabel ?? "items")
        return "\(count) \(noun)"
    }

    public static func permissionAllows(
        spec: PermissionBoundarySpec,
        authorization: [String: JSONValue],
        rows: [[String: JSONValue]] = [],
        grants: [[String: JSONValue]] = []
    ) -> Bool {
        guard let capability = spec.capability?.trimmingCharacters(in: .whitespacesAndNewlines), !capability.isEmpty else { return true }
        let mode = spec.mode?.trimmingCharacters(in: .whitespacesAndNewlines).lowercased() ?? "resource"
        if mode == "resource" {
            return authorization["resource"]?.objectValue?["capabilities"]?.objectValue?[capability]?.boolValue == true
        }
        guard !rows.isEmpty else { return true }
        let identity = spec.identityField?.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty == false ? spec.identityField! : "id"
        let grantsByID = Dictionary(uniqueKeysWithValues: grants.compactMap { grant -> (String, [String: JSONValue])? in
            guard let value = grant["resourceId"] ?? grant[identity], let key = primitiveIdentityString(value) else { return nil }
            return (key, grant)
        })
        return rows.allSatisfy { row in
            guard let value = row[identity], let key = primitiveIdentityString(value), let grant = grantsByID[key] else { return false }
            return grant["capabilities"]?.objectValue?[capability]?.boolValue == true
        }
    }

    public static func responsiveDataGridState(spec: ResponsiveDataGridSpec, target: String) -> ResponsiveDataGridStateSpec? {
        let states = spec.breakpoints ?? [:]
        return states[target] ?? states[target == "phone" ? "narrow" : "wide"] ?? states["desktop"] ?? states.values.first
    }

    public static func validateSchedule(rows: [[String: JSONValue]], spec: ScheduleEditorSpec) -> ScheduleValidationResult {
        let startField = spec.startField ?? "start"
        let endField = spec.endField ?? "end"
        let minimum = parseDurationMilliseconds(spec.minDuration)
        var normalized: [(index: Int, start: Date?, end: Date?)] = []
        var errors: [ScheduleValidationError] = []
        for (index, row) in rows.enumerated() {
            if row["_scheduleErrors"]?.objectValue?.values.contains(where: { $0.boolValue == true || $0.stringValue?.isEmpty == false }) == true {
                errors.append(ScheduleValidationError(index: index, code: "timezone"))
            }
            let start = scheduleDate(resolveSchedule(row, selector: startField))
            let end = scheduleDate(resolveSchedule(row, selector: endField))
            normalized.append((index, start, end))
            guard let start, let end else {
                errors.append(ScheduleValidationError(index: index, code: "invalid_date"))
                continue
            }
            if end <= start { errors.append(ScheduleValidationError(index: index, code: "invalid_range")) }
            if let minimum, end.timeIntervalSince(start) * 1_000 < minimum {
                errors.append(ScheduleValidationError(index: index, code: "min_duration"))
            }
        }
        if spec.allowOverlap == false {
            let sorted = normalized.compactMap { item -> (Int, Date, Date)? in
                guard let start = item.start, let end = item.end else { return nil }
                return (item.index, start, end)
            }.sorted { $0.1 < $1.1 }
            for index in 1..<sorted.count where sorted[index].1 < sorted[index - 1].2 {
                errors.append(ScheduleValidationError(index: sorted[index].0, code: "overlap"))
            }
        }
        return ScheduleValidationResult(valid: errors.isEmpty, errors: errors)
    }

    public static func toggleTreeSelection(
        nodes: [[String: JSONValue]],
        selected: Set<String>,
        key: String,
        checked: Bool,
        spec: TreeEditorSpec
    ) -> Set<String> {
        var result = selected
        var keys: Set<String> = [key]
        if spec.cascade?.lowercased() == "descendants" {
            for node in nodes { collectTreeSelectionKeys(node: node, target: key, collecting: false, spec: spec, result: &keys) }
        }
        for value in keys { if checked { result.insert(value) } else { result.remove(value) } }
        return result
    }

    public static func validateUpload(files: [UploadFileValue], spec: UploadCollectionSpec) -> UploadValidationResult {
        var errors: [String] = []
        if spec.multiple == false && files.count > 1 { errors.append("Only one file is allowed.") }
        if let maxFiles = spec.maxFiles, maxFiles > 0, files.count > maxFiles { errors.append("No more than \(maxFiles) files are allowed.") }
        for file in files {
            if !uploadTypeAccepted(file, rules: spec.accept ?? []) { errors.append("\(file.name) has an unsupported type.") }
            if let maxBytes = spec.maxBytes, maxBytes > 0, Int64(file.data.count) > maxBytes { errors.append("\(file.name) exceeds the size limit.") }
        }
        return UploadValidationResult(valid: errors.isEmpty, errors: errors)
    }

    public static func encodeUpload(files: [UploadFileValue], spec: UploadCollectionSpec) throws -> [String: JSONValue] {
        let transport = spec.transport ?? "mcpBlob"
        guard transport == "mcpBlob" else { throw UploadRuntimeError.message("Unsupported upload transport: \(transport)") }
        let blobs = files.map { file in JSONValue.object([
            "name": .string(file.name), "size": .number(Double(file.data.count)), "type": .string(file.mimeType),
            "data": .string(file.data.base64EncodedString()), "filename": .string(file.name), "mimeType": .string(file.mimeType)
        ]) }
        return [
            "transport": .string("mcpBlob"),
            spec.blobField ?? "files": spec.multiple == false ? (blobs.first ?? .null) : .array(blobs)
        ]
    }

    public static func masterDetailIdentity(row: [String: JSONValue]?, fields: [String]) -> String? {
        guard let row else { return nil }
        let keys = fields.isEmpty ? ["id"] : fields
        let values = keys.compactMap { row[$0].flatMap(primitiveIdentityString) }
        return values.count == keys.count ? values.joined(separator: "\u{001f}") : nil
    }

    public static func resolveMasterDetailSelection(
        rows: [[String: JSONValue]],
        selected: [String: JSONValue]?,
        persisted: [String: JSONValue]?,
        identityFields: [String]
    ) -> [String: JSONValue]? {
        let selectedID = masterDetailIdentity(row: selected, fields: identityFields)
        if let selectedID, let match = rows.first(where: { masterDetailIdentity(row: $0, fields: identityFields) == selectedID }) { return match }
        let persistedID = masterDetailIdentity(row: persisted, fields: identityFields)
        if let persistedID { return rows.first(where: { masterDetailIdentity(row: $0, fields: identityFields) == persistedID }) }
        return nil
    }

    public static func masterDetailParameters(spec: MasterDetailRegionSpec, row: [String: JSONValue]) -> [String: JSONValue] {
        spec.parameters.compactMapValues { parameter in
            guard let object = parameter.objectValue else { return parameter }
            let source = object["source"]?.stringValue?.lowercased() ?? "row"
            guard source == "row" else { return object["value"] ?? parameter }
            let selector = object["selector"]?.stringValue ?? object["field"]?.stringValue ?? ""
            let value = resolveSchedule(row, selector: selector) ?? .null
            return object["wrap"]?.stringValue?.lowercased() == "array" ? .array([value]) : value
        }
    }

    private static func collectTreeSelectionKeys(node: [String: JSONValue], target: String, collecting: Bool, spec: TreeEditorSpec, result: inout Set<String>) {
        let identityField = spec.identityField ?? "id"
        let nodeKey = node[identityField].flatMap(primitiveIdentityString) ?? ""
        let shouldCollect = collecting || nodeKey == target
        if shouldCollect && !nodeKey.isEmpty { result.insert(nodeKey) }
        let children = node[spec.childrenField ?? "children"]?.arrayValue?.compactMap(\.objectValue) ?? []
        for child in children { collectTreeSelectionKeys(node: child, target: target, collecting: shouldCollect, spec: spec, result: &result) }
    }

    public static func historyDiffEntries(before: JSONValue?, after: JSONValue?, spec: HistoryDiffSpec) -> [HistoryDiffEntry] {
        var result: [HistoryDiffEntry] = []
        collectHistoryDiff(before: before ?? .null, after: after ?? .null, path: "", spec: spec, result: &result)
        return result
    }

    private static func collectHistoryDiff(before: JSONValue, after: JSONValue, path: String, spec: HistoryDiffSpec, result: inout [HistoryDiffEntry]) {
        let leaf = path.split(separator: ".").last.map(String.init) ?? path
        let ignored = Set(spec.ignoreFields ?? [])
        if ignored.contains(path) || ignored.contains(leaf) { return }
        if case .object(let left) = before, case .object(let right) = after {
            for key in Set(left.keys).union(right.keys).sorted() {
                collectHistoryDiff(
                    before: left[key] ?? .null,
                    after: right[key] ?? .null,
                    path: path.isEmpty ? key : "\(path).\(key)",
                    spec: spec,
                    result: &result
                )
            }
            return
        }
        let normalizedBefore = normalizedHistoryValue(before, arrayStrategy: spec.arrayStrategy)
        let normalizedAfter = normalizedHistoryValue(after, arrayStrategy: spec.arrayStrategy)
        guard normalizedBefore != normalizedAfter else { return }
        let redactedFields = Set(spec.redactFields ?? [])
        let redacted = redactedFields.contains(path) || redactedFields.contains(leaf)
        result.append(HistoryDiffEntry(
            path: path.isEmpty ? "value" : path,
            label: spec.fieldLabels?[path] ?? spec.fieldLabels?[leaf] ?? historyLabel(leaf),
            before: redacted ? "••••" : historyDisplay(before),
            after: redacted ? "••••" : historyDisplay(after),
            redacted: redacted
        ))
    }

    private static func normalizedHistoryValue(_ value: JSONValue, arrayStrategy: String?) -> JSONValue {
        guard arrayStrategy?.lowercased() == "set", case .array(let values) = value else { return value }
        return .array(values.sorted { historyDisplay($0) < historyDisplay($1) })
    }

    private static func historyDisplay(_ value: JSONValue) -> String {
        switch value {
        case .string(let value): return value
        case .number(let value): return value.rounded() == value ? String(Int(value)) : String(value)
        case .bool(let value): return value ? "true" : "false"
        case .null: return "—"
        case .array(let values): return values.map(historyDisplay).joined(separator: ", ")
        case .object(let values): return values.keys.sorted().map { "\($0): \(historyDisplay(values[$0] ?? .null))" }.joined(separator: ", ")
        }
    }

    private static func historyLabel(_ value: String) -> String {
        value.replacingOccurrences(of: "_", with: " ").split(separator: " ").map { $0.prefix(1).uppercased() + $0.dropFirst() }.joined(separator: " ")
    }
}

private enum UploadRuntimeError: LocalizedError {
    case message(String)
    var errorDescription: String? { if case .message(let value) = self { return value }; return nil }
}

private func uploadTypeAccepted(_ file: UploadFileValue, rules: [String]) -> Bool {
    if rules.isEmpty { return true }
    let name = file.name.lowercased(); let type = file.mimeType.lowercased()
    return rules.contains { raw in
        let rule = raw.lowercased()
        if rule.hasPrefix(".") { return name.hasSuffix(rule) }
        if rule.hasSuffix("/*") { return type.hasPrefix(String(rule.dropLast())) }
        return type == rule
    }
}

private func resolveSchedule(_ row: [String: JSONValue], selector: String) -> JSONValue? {
    JSONValue(any: SelectorUtil.resolve(row.compactMapValues(\.anyValue), selector: selector))
}

private func scheduleDate(_ value: JSONValue?) -> Date? {
    guard case .string(let text) = value else { return nil }
    let formatter = ISO8601DateFormatter()
    formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
    return formatter.date(from: text) ?? ISO8601DateFormatter().date(from: text)
}

private func parseDurationMilliseconds(_ raw: String?) -> Double? {
    let text = raw?.trimmingCharacters(in: .whitespacesAndNewlines).lowercased() ?? ""
    guard let match = text.range(of: #"^(\d+(?:\.\d+)?)\s*(ms|s|m|h|d)$"#, options: .regularExpression) else { return nil }
    let matched = String(text[match])
    let unit = matched.hasSuffix("ms") ? "ms" : String(matched.suffix(1))
    let numberText = unit == "ms" ? String(matched.dropLast(2)) : String(matched.dropLast())
    guard let value = Double(numberText.trimmingCharacters(in: .whitespaces)) else { return nil }
    let multiplier: Double = ["ms": 1, "s": 1_000, "m": 60_000, "h": 3_600_000, "d": 86_400_000][unit] ?? 1
    return value * multiplier
}

private func primitiveIdentityString(_ value: JSONValue) -> String? {
    switch value {
    case .string(let value): return value.isEmpty ? nil : value
    case .number(let value): return value.rounded() == value ? String(Int(value)) : String(value)
    case .bool(let value): return value ? "true" : "false"
    default: return nil
    }
}
